### PR TITLE
701 restrict derivation

### LINKED ISSUE #701: Do we need a `slice_derivation()` function?
Do we need a function which splits the input dataset in two or more parts and calls different derivations for these parts?

This was requested for
- `CHG` and `PCHG` where the variables should not be derived for baseline or before baseline (see https://github.com/Roche-GSK/admiral/issues/676)
- imputations where different imputation rules should be used depending on a condition (see https://github.com/Roche-GSK/admiral/issues/686)

Some functions provide already a similar functionality, e.g., `derive_worst_flag()` and `derive_extreme_flag()`.

**Decision from Questions and Comments meeting**

- Implement `restrict_derivation(dataset, derivation, args, filter)`. If the derivation needs to be applied to more than one subset of the input dataset, multiple `restrict_derivation()` calls should be chained, e.g.,
  ```
  restrict_derivation(
    dataset = adae ,
    derivation = derive_vars_dtm ,
    args = params(
      dtc = AEENDTC,
      new_vars_prefix = "AEN",
      date_imputation = "last",
      time_imputation = "last",
      max_dates = vars(DTHDT)),
    filter = AEOUT == "FATAL"
  ) %>%
  restrict_derivation(
    derivation = derive_vars_dtm ,
    args = params(
      dtc = AEENDTC,
      new_vars_prefix = "AEN",
      date_imputation = "last",
      time_imputation = "last"),
    filter = AEOUT != "FATAL" or is.na(AEOUT)
  )
  ```
- remove `filter` argument from functions if possible
- add examples for `restrict_derivation()` to functions which are commonly used with `restrict_derivation()`, e.g., `derive_var_chg()`, `derive_var_pchg()`
- update documentation regarding definition of derivations (it is acceptable to split a derivation into multiple function calls) and add some examples

### PR DESCRIPTION
**Pull request by bundfussr**
Closes #701 

Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/) 
- [x] Updated relevant unit tests or have written new unit tests - See [Unit Test Guide](https://pharmaverse.github.io/admiral/articles/unit_test_guidance.html#writing-unit-tests-in-admiral-)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiral/articles/programming_strategy.html#deprecation-1)?
- [x] Update to all relevant roxygen headers and examples 
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md`
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue so that it closes after successful merging. 
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done!  Much love to your accomplishment!

### CODE REVIEW

**Comment by bundfussr** [comment]
@pharmaverse/admiral , would anyone like to review this pull request? It implements `restrict_derivation()` and `slice_derivation()`.

I will solve the issues in the R-CMD checks when I am back from vacation (13th to 20th April).

File: NEWS.md
Comment (millerg23) [inline NEWS.md:37]: Think lione 37 should be `The input dataset is split into slices (subsets) and`

File: R/restrict_derivation.R
Comment (millerg23) [inline R/restrict_derivation.R:3]: Remove line 3, its a repeat of Line 1.

File: R/slice_derivation.R
Comment (millerg23) [inline R/slice_derivation.R:3]: Think line 3 should read `The input dataset is split into slices (subsets) and for each slice the`

File: R/slice_derivation.R
Comment (millerg23) [inline R/slice_derivation.R:18]: Typo in line 18 `or` should be `for`

File: R/slice_derivation.R
Comment (millerg23) [inline R/slice_derivation.R:31]: line 31 I think could be clearer with`Observations that match with more than one slice are only considered for the first matching slice`

**Comment by millerg23** [comment]
Finshed review, looks great! Just some minor comments. Compared most of the code to what was done in `call_derivation`, and see no issues. Once `DEVEL` branch merged in, can re-check everything with regard to building website and reviewing pages. Ideally @thomas-neitmann could have a quick look over the functions just for sanity check.

File: NEWS.md
Comment (bundfussr) [reply NEWS.md:37]: Updated

File: R/restrict_derivation.R
Comment (bundfussr) [reply R/restrict_derivation.R:3]: If I remove line 3, the line is repeated in the rendered documentation but both lines are in title case. This looks a bit odd. Thus I would keep it.

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:3]: Updated

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:18]: Fixed

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:31]: Updated

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:34]: Updated

**Comment by bundfussr** [comment]
@millerg23 , I have added a work-around to avoid that the `slice_derivation()` unit tests fail in R 3.6.

The issue in the R-CMD check using the latest version should resolve once we merge https://github.com/pharmaverse/admiral/pull/1078.

File: R/slice_derivation.R
Comment (thomas-neitmann) [inline R/slice_derivation.R:118]:
```
dataset_split %>%
  unnest() %>%
  ungroup() %>%
  select(-temp_slicenr)
```

File: _pkgdown.yml
Comment (thomas-neitmann) [inline _pkgdown.yml:72]: There should be no `,` after "e.g.".

File: R/derive_var_extreme_flag.R
Comment (thomas-neitmann) [inline R/derive_var_extreme_flag.R:363]: Function name should be in backticks. Also, it should be rest**r**ict.

File: R/slice_derivation.R
Comment (thomas-neitmann) [reply R/slice_derivation.R:144]: Also not quite sure if we should have default values. It looks a bit odd to me in the example of `slice_derivation()` that for the second slice `filter` is not specified. I'd personally prefer if both parameters would have to be set explicitly.

File: R/slice_derivation.R
Comment (thomas-neitmann) [inline R/slice_derivation.R:58]: Can we make this somehow more explicit? I feel like just not specifiying the `filter` parameter is confusing...

**Review by thomas-neitmann — CHANGES_REQUESTED** [review]

File: R/restrict_derivation.R
Comment (bundfussr) [reply R/restrict_derivation.R:38]: Updated

File: R/restrict_derivation.R
Comment (bundfussr) [reply R/restrict_derivation.R:84]: Using `data_derive`

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:121]: Fixed

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:144]: I have no strong opinion. So I removed the default values.

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:118]: updated

File: _pkgdown.yml
Comment (bundfussr) [reply _pkgdown.yml:72]: I think it depends on the style guide. It seems that American style guides prefer the comma while British one doesn't.

I have removed it.

File: R/derive_var_extreme_flag.R
Comment (bundfussr) [reply R/derive_var_extreme_flag.R:363]: Fixed

File: R/slice_derivation.R
Comment (bundfussr) [reply R/slice_derivation.R:58]: Updated

**Review by millerg23 — COMMENTED** [review]
No further comments from me @bundfussr , nice job. Just need to fix merge conflicts. Will leave @thomas-neitmann to approve, or Thomas let me know and I can approve if you are happy with changes made.

**Comment by bundfussr** [comment]
@millerg23 , @thomas-neitmann , I fixed the merge conflicts.

**Review by thomas-neitmann — APPROVED** [review]
Great work 👏

**Comment by bundfussr** [comment]
@thomas-neitmann , could you merge the PR (before new merge conflicts come up)? Or is there anything outstanding?

**Pull request merged**
